pkg-manager/plugin-commands-installation/test/__snapshots__/dedupe.ts.snap

```diff
@@ -0,0 +1,197 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`removes packages from pnpm-lock.yaml 1`] = `
+{
+  "importers": {
+    ".": {
+      "specifiers": {},
+    },
+    "packages/bar": {
+      "dependencies": {
+        "ajv": "6.10.2",
+      },
+      "specifiers": {
+        "ajv": "^6.10.2",
+      },
+    },
+    "packages/foo": {
+      "dependencies": {
+        "ajv": "6.12.6",
+      },
+      "specifiers": {
+        "ajv": "^6.12.5",
+      },
+    },
+  },
+  "lockfileVersion": 5.4,
+  "packages": {
+    "/ajv/6.10.2": {
+      "dependencies": {
+        "fast-deep-equal": "2.0.1",
+        "fast-json-stable-stringify": "2.0.0",
+        "json-schema-traverse": "0.4.1",
+        "uri-js": "4.2.2",
+      },
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-TXtUUEYHuaTEbLZWIKUr5pmBuhDLy+8KYtPYdcV8qC+pOZL+NKqYwvWSRrVXHn+ZmRRAu8vJTAznH7Oag6RVRw==",
+      },
+    },
+    "/ajv/6.12.6": {
+      "dependencies": {
+        "fast-deep-equal": "3.1.3",
+        "fast-json-stable-stringify": "2.1.0",
+        "json-schema-traverse": "0.4.1",
+        "uri-js": "4.4.1",
+      },
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-j3fVLgvTo527anyYyJOGTYJbG+vnnQYvE0m5mmkc1TK+nxAppkCLMIL0aZ4dblVCNoGShhm+kzE4ZUykBoMg4g==",
+      },
+    },
+    "/fast-deep-equal/2.0.1": {
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-bCK/2Z4zLidyB4ReuIsvALH6w31YfAQDmXMqMx6FyfHqvBxtjC0eRumeSu4Bs3XtXwpyIywtSTrVT99BxY1f9w==",
+      },
+    },
+    "/fast-deep-equal/3.1.3": {
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-f3qQ9oQy9j2AhBe/H9VC91wLmKBCCU/gDOnKNAYG5hswO7BLKj09Hc5HYNz9cGI++xlpDCIgDaitVs03ATR84Q==",
+      },
+    },
+    "/fast-json-stable-stringify/2.0.0": {
+      "dev": false,
+      "resolution": {
+        "integrity": "sha1-1RQsDK7msRifh9OnYREGT4bIu/I=",
+      },
+    },
+    "/fast-json-stable-stringify/2.1.0": {
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-lhd/wF+Lk98HZoTCtlVraHtfh5XYijIjalXck7saUtuanSDyLMxnHhSXEDJqHxD7msR8D0uCmqlkwjCV8xvwHw==",
+      },
+    },
+    "/json-schema-traverse/0.4.1": {
+      "dev": false,
+      "resolution": {
+        "integrity": "sha512-xbbCH5dCYU5T8LcEhhuh7HJ88HXuW3qsI3Y0zOZFKfZEHcpWiHU/Jxzk629Brsab/mMiHQti9wMP+845RPe3Vg==",
+      },
+    },
+    "/punycode/2.1.1": {
+      "dev": false,
+      "engines": {
+        "node": ">=6",
+      },
+      "resolution": {
+        "integrity": "sha512-XRsRjdf+j5ml+y/6GKHPZbrF/8p2Yga0JPtdqTIY2Xe5ohJPD9saDJJLPvp9+NSBprVvevdXZybnj2cv8OEd0A==",
+      },
+    },
+    "/punycode/2.3.0": {
```

Comment:
Is it expected that two versions of `punycode` are present? And both are v2

Reply:
Yeah, the other version comes in from the earlier version of `uri-js@4.2.2`, which also appears twice in the lockfile before deduping.

```
❯ pnpm view uri-js@4.2.2

# Omitted for brevity

dependencies:
punycode: ^2.1.0 
```

The earlier version of `uri-js@4.2.2` is brought in by the earlier version of `ajv`.

```
❯ pnpm view ajv@6.10.2  

# Omitted for brevity

dependencies:
fast-deep-equal: ^2.0.1
fast-json-stable-stringify: ^2.0.0
json-schema-traverse: ^0.4.1
uri-js: ^4.2.2
```

Fortunately there's only one version of `uri-js` and `punycode` after deduping. 🙂

Reply:
It might be helpful to test that if `uri-js@4.2.2` and `uri-js@4.4.1` are present in the lockfile, that `uri-js@4.2.2` dedupes to use `punycode@2.3.0`. I'll add that test now.

Reply:
I could clarify one more thing here: The earlier versions of `ajv`'s dependencies appear because I copied the `with-peer` fixture.

https://github.com/pnpm/pnpm/blob/1072ec12869fcad320459754b5fbbac09d7787a0/__fixtures__/with-peer/package.json#L2-L5

I then added a new workspace package that depends on the later version of `ajv` of that same major version. That should simulate what happens in a lot of monorepos, where a developer adds a later version of a dependency that already existed.

Reply:
> It might be helpful to test that if `uri-js@4.2.2` and `uri-js@4.4.1` are present in the lockfile, that `uri-js@4.2.2` dedupes to use `punycode@2.3.0`. I'll add that test now.

Done in https://github.com/pnpm/pnpm/pull/5958/commits/8010d6165a1e33913fbed324900d9e8e53f47f51